DataDog/dd-trace-dotnet
# Handle client-disconnect scenario in IIS

## Summary of changes

Handle case where IIS returns a `400` response but reports a `500` response

## Reason for change

Fixes a client-disconnect scenario, where the client terminates a connection while the server is waiting for it to send data. In these scenarios, IIS effectively returns a `400 Bad Request`, but it _reports_ a `500` error to the aspnet pipeline. Consequently, we "incorrectly" record `500` errors on spans.

## Implementation details

Check for the specific client disconnect scenario in `OnEndRequest` and change the response to `400` in this case.

## Test coverage

Added integration tests to test this scenario using manual sockets:
1. Open a socket
2. Send a `GET` request with an `Expect: 100-continue` header
3. The server responds `HTTP/1.1 100 Continue`, and expects the client to send more data
4. Disconnect

The first commit demonstrates this scenario and the snapshots as it happens now. You can see the `500` response codes (in the cases where we generate snapshots). The second commit fixes the issue, and shows how this updates the snapshots.

Some extra notes on the tests
- I only added this to a single IIS test, as it seems unnecessary to test all versions of aspnet when this is essentially an IIS behaviour
- Classic mode doesn't generate _any_ spans, so can't really test this behaviour. Kept it in the tests to ensure we don't crash in this case.
- I tried to test this when using aspnetcore with the IIS proxy, but it doesn't work. When running out-of-process, IIS returns the `HTTP/1.1 100` response, but ASP.NET Core returns a 200 response. When running in-process, ASP.NET Core responds `200 OK`, and so there's _no_ `HTTP/1.1 100` response at all, which is weird... Ignored it for now anyway as not sure what to make of it, and the original issue was netframework... we can revisit if it's a netcore issue later I think
